### ISSUE TITLE
Extract single frame from video 

### ISSUE THREAD

**Issue by Piyush-k-singh** [ultralytics/yolov5#7195]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

Extract single frame from video which has higher confidence score. No need to crop all the frame which is detected in the video. 

### Use case

Many times we will get video as an input from that video we need to extract one frame then we cropped that frame and do our analysis. what more feature we can extract in that frame.

### Additional

I am not able to cropped the best frame from video if it has higher confidence score.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by Piyush-k-singh** [comment]
Finally,  I build build this 👍.

Input as video and extract one frame from the video based on confidence level.

**Comment by ashrithjacob** [comment]
Hi Piyush,
Where can I get this feature?

**Comment by syekysushe** [comment]
Hi @Piyush-k-singh ,

Can you guide us on how you do it? Really need some assistance on it for my project right now. Thank you.

**Comment by glenn-jocher** [comment]
Hello @syekysushe, glad to hear that you are interested in this feature! As of now, the feature requested by @Piyush-k-singh is not yet available in YOLOv5. However, @Piyush-k-singh mentioned that he was able to build the feature. 

@Piyush-k-singh, Can you please share some more details on how you built this feature? It would be great to have some guidance on how to implement this in YOLOv5. Thank you!

**Comment by Piyush-k-singh** [comment]
while detection in video you can check conf. score of each frame and compare current frame with previous frame based on conf. score  if current frame conf. > previous frame conf. then save that frame or crop that object.
it will only work if your yolo detection performing well.

**Comment by syekysushe** [comment]
Hello @Piyush-k-singh , 

Can you show a snippet of code where I could add this feature to?

**Comment by Piyush-k-singh** [comment]
@glenn-jocher if you need the code i need to check where it is or i have to write again because i have done that for someone else.
but i use above technique to solve that problem.

**Comment by Piyush-k-singh** [comment]
# Write results
                for *xyxy, conf, cls in reversed(det):
                    if save_txt:  # Write to file
                        xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                        line = (cls, *xywh, conf) if save_conf else (cls, *xywh)  # label format
                        with open(f'{txt_path}.txt', 'a') as f:
                            f.write(('%g ' * len(line)).rstrip() % line + '\n')

                    if save_img or save_crop or view_img:  # Add bbox to image
                        c = int(cls)  # integer class
                        label = None if hide_labels else (names[c] if hide_conf else f'{names[c]} {conf:.2f}')
                        annotator.box_label(xyxy, label, color=colors(c, True))
                    if save_crop:
                        save_one_box(xyxy, imc, file=save_dir / 'crops' / names[c] / f'{p.stem}.jpg', BGR=True)**

In detect.py  you need play here to get one frame from video based on conf. score
I am searching for code where it is if i'll find that i will share with you all.

**Comment by glenn-jocher** [comment]
Hello @Piyush-k-singh, 

Thank you for your interest in YOLOv5! Based on the information provided by @Piyush-k-singh, the code snippet provided is from the `detect.py` file in YOLOv5. According to @Piyush-k-singh, you might need to make changes in the `detect.py` file to extract one frame from the video based on confidence score. 

To get one frame from a video based on confidence score, you can check the confidence score of each frame and compare the current frame with the previous frame based on the confidence score. If the confidence score of the current frame is greater than the confidence score of the previous frame, then you can save that frame or crop the object in that frame. However, please note that this method will only work if your YOLO detection is performing well.

If you need further assistance, feel free to ask and we will do our best to help you.

**Comment by syekysushe** [comment]
Hello @Piyush-k-singh ,

Thank you so much for the guide. Please share with me how you did the code once you found it. I am still a bit stuck at this part. Again. thank you.

**Comment by glenn-jocher** [comment]
Hello @syekysushe,

I'm glad that @Piyush-k-singh's guidance was helpful to you. Unfortunately, @Piyush-k-singh hasn't shared the code for this feature yet. However, as @Piyush-k-singh mentioned, you can implement this feature by checking the confidence score of each frame and comparing it with the confidence score of the previous frame. If the current frame's confidence score is higher, you can save the frame or crop the object in that frame.

If you have any further questions or need additional assistance, please don't hesitate to ask. We're here to help.

**Comment by syekysushe** [comment]
Hello again,

I have managed to solve this problem. Below is the code I edited in detect.py if anyone is facing the same issue. The items in ** are ones I added (remove the ** if you want to copy and paste the code). I save a cropped image from a frame that has the largest area of bounding box (I check this first) and then highest value of confidence level throughout all frames.

```
   # Run inference
    model.warmup(imgsz=(1 if pt or model.triton else bs, 3, *imgsz))  # warmup
    seen, windows, dt = 0, [], (Profile(), Profile(), Profile())

    **max_conf = 0
    max_area = 0**
    
    **# Define a variable of type *xyxy
    xyxy_max: Tuple[float, float, float, float] = (0.0, 0.0, 0.0, 0.0)**

    for path, im, im0s, vid_cap, s in dataset:
        with dt[0]:
            im = torch.from_numpy(im).to(model.device)
            im = im.half() if model.fp16 else im.float()  # uint8 to fp16/32
            im /= 255  # 0 - 255 to 0.0 - 1.0
            if len(im.shape) == 3:
                im = im[None]  # expand for batch dim

        # Inference
        with dt[1]:
            visualize = increment_path(save_dir / Path(path).stem, mkdir=True) if visualize else False
            pred = model(im, augment=augment, visualize=visualize)

        # NMS
        with dt[2]:
            pred = non_max_suppression(pred, conf_thres, iou_thres, classes, agnostic_nms, max_det=max_det)

        # Second-stage classifier (optional)
        # pred = utils.general.apply_classifier(pred, classifier_model, im, im0s)

        # Process predictions
        for i, det in enumerate(pred):  # per image
            seen += 1
            if webcam:  # batch_size >= 1
                p, im0, frame = path[i], im0s[i].copy(), dataset.count
                s += f'{i}: '
            else:
                p, im0, frame = path, im0s.copy(), getattr(dataset, 'frame', 0)

            p = Path(p)  # to Path
            save_path = str(save_dir / p.name)  # im.jpg
            txt_path = str(save_dir / 'labels' / p.stem) + ('' if dataset.mode == 'image' else f'_{frame}')  # im.txt
            s += '%gx%g ' % im.shape[2:]  # print string
            gn = torch.tensor(im0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
            # imc = im0.copy() if save_crop else im0  # for save_crop
            annotator = Annotator(im0, line_width=line_thickness, example=str(names))
            if len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()

                # Print results
                for c in det[:, 5].unique():
                    n = (det[:, 5] == c).sum()  # detections per class
                    s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string

                # Write results
                for *xyxy, conf, cls in reversed(det):

                    if save_txt:  # Write to file
                        xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                        line = (cls, *xywh, conf) if save_conf else (cls, *xywh)  # label format
                        with open(f'{txt_path}.txt', 'a') as f:
                            f.write(('%g ' * len(line)).rstrip() % line + '\n')

                    if save_img or save_crop or view_img:  # Add bbox to image
                        c = int(cls)  # integer class
                        label = None if hide_labels else (names[c] if hide_conf else f'{names[c]} {conf:.2f}')
                        annotator.box_label(xyxy, label, color=colors(c, True))
                    if save_crop:
                        **print("XYWH 2 WIDTH = ", xywh[2])
                        print("XYWH 3, HEIGHT = ", xywh[3])
                        conf_int = conf.item()
                        print("CONF = ", conf_int)
                        area = xywh[2] * xywh [3]
                        if area >= max_area:
                            if conf_int >= max_conf:
                                max_area = area
                                max_conf = conf
                                max_conf_frame = im0.copy()
                                xyxy_max = xyxy**

            # Stream results
            im0 = annotator.result()
            if view_img:
                if platform.system() == 'Linux' and p not in windows:
                    windows.append(p)
                    cv2.namedWindow(str(p), cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO)  # allow window resize (Linux)
                    cv2.resizeWindow(str(p), im0.shape[1], im0.shape[0])
                cv2.imshow(str(p), im0)
                cv2.waitKey(1)  # 1 millisecond

            # Save results (image with detections)
            if save_img:
                if dataset.mode == 'image':
                    cv2.imwrite(save_path, im0)
                else:  # 'video' or 'stream'
                    if vid_path[i] != save_path:  # new video
                        vid_path[i] = save_path
                        if isinstance(vid_writer[i], cv2.VideoWriter):
                            vid_writer[i].release()  # release previous video writer
                        if vid_cap:  # video
                            fps = vid_cap.get(cv2.CAP_PROP_FPS)
                            w = int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                            h = int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                        else:  # stream
                            fps, w, h = 30, im0.shape[1], im0.shape[0]
                        save_path = str(Path(save_path).with_suffix('.mp4'))  # force *.mp4 suffix on results videos
                        vid_writer[i] = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*'mp4v'), fps, (w, h))
                    vid_writer[i].write(im0)

        # Print time (inference-only)
        LOGGER.info(f"{s}{'' if len(det) else '(no detections), '}{dt[1].dt * 1E3:.1f}ms")
    
    **cv2.imwrite(os.path.join(save_dir/"fullframe.jpg"), max_conf_frame)
    save_one_box(xyxy_max, max_conf_frame, file=save_dir / 'cropped.jpg', BGR=True)**
```

**Comment by glenn-jocher** [comment]
@syekysushe hello!

Thank you for sharing the solution code to this issue. We really appreciate your contribution and we believe it will be helpful to others who might have a similar problem.

If you have any other questions or concerns, please do not hesitate to ask. We are always here to help.
